File: pages/auth.page.js
/**
 * Authentication Page Object (Login & Register)
 */

const { By } = require('selenium-webdriver');
const BasePage = require('./base.page');
const logger = require('../utilities/logger');

class AuthPage extends BasePage {
  constructor(driver) {
    super(driver);

    // Form Locators
    this.LOGIN_TAB = By.css('button[data-tab="login"], a[href="#login"], #tab-login');
    this.REGISTER_TAB = By.css('button[data-tab="register"], a[href="#register"], #tab-register');
    this.USERNAME_INPUT = By.css('input[name="username"], #username, input[type="text"]');
    this.PASSWORD_INPUT = By.css('input[name="password"], #password, input[type="password"]');
    this.CONFIRM_PASSWORD_INPUT = By.css('input[name="confirm_password"], #confirm_password');
    this.SUBMIT_BTN = By.css('button[type="submit"], input[type="submit"], .auth-submit-btn');
    this.ERROR_MESSAGE = By.css('.error, .alert-danger, .flash-message.error, .auth-error, [role="alert"]');
    this.SUCCESS_MESSAGE = By.css('.success, .alert-success, .flash-message.success');
    this.GOOGLE_LOGIN_BTN = By.css('a[href*="/login/google"], .google-login-btn');
    this.AUTH_CONTAINER = By.css('.auth-card, .auth-container, .login-form');
  }

  async openLogin() {
    await this.navigateTo('/login');
    return this;
  }

  async openRegister() {
    await this.navigateTo('/register');
    return this;
  }

  async switchToLoginTab() {
    if (await this.isElementVisible(this.LOGIN_TAB)) {
      await this.clickElement(this.LOGIN_TAB);
    }
  }

  async switchToRegisterTab() {
    if (await this.isElementVisible(this.REGISTER_TAB)) {
      await this.clickElement(this.REGISTER_TAB);
    }
  }

  async login(username, password) {
    logger.info(`[AuthPage] Attempting login with username: "${username}"`);
    if (username !== null && username !== undefined) {
      await this.typeInto(this.USERNAME_INPUT, username);
    }
    if (password !== null && password !== undefined) {
      await this.typeInto(this.PASSWORD_INPUT, password);
    }
    await this.clickElement(this.SUBMIT_BTN);
    await this.utils.sleep(500);
  }

  async register(username, password, confirmPassword = password) {
    logger.info(`[AuthPage] Attempting registration for username: "${username}"`);
    await this.switchToRegisterTab();
    if (username) await this.typeInto(this.USERNAME_INPUT, username);
    if (password) await this.typeInto(this.PASSWORD_INPUT, password);
    if (await this.isElementVisible(this.CONFIRM_PASSWORD_INPUT) && confirmPassword) {
      await this.typeInto(this.CONFIRM_PASSWORD_INPUT, confirmPassword);
    }
    await this.clickElement(this.SUBMIT_BTN);
    await this.utils.sleep(500);
  }

  async getErrorMessage() {
    if (await this.isElementVisible(this.ERROR_MESSAGE)) {
      return await this.getElementText(this.ERROR_MESSAGE);
    }
    return '';
  }

  async isAuthCardVisible() {
    return await this.isElementVisible(this.AUTH_CONTAINER);
  }
}

module.exports = AuthPage;

File: tests/pages/auth_page.py
"""
Page Object for Authentication (Login and Registration).
"""

from selenium.webdriver.common.by import By
from tests.pages.base_page import BasePage


class AuthPage(BasePage):
    """Page Object covering /login and /register flows."""

    # Selectors
    WORDMARK = (By.CSS_SELECTOR, "[data-testid='auth-wordmark']")
    TITLE = (By.CSS_SELECTOR, "[data-testid='auth-title']")
    EMAIL_INPUT = (By.CSS_SELECTOR, "[data-testid='email-input']")
    PASSWORD_INPUT = (By.CSS_SELECTOR, "[data-testid='password-input']")
    CONFIRM_PASSWORD_INPUT = (By.CSS_SELECTOR, "[data-testid='confirm-password-input']")
    SUBMIT_BTN = (By.CSS_SELECTOR, "[data-testid='auth-submit-btn']")
    GOOGLE_BTN = (By.CSS_SELECTOR, "[data-testid='google-login-btn']")
    FLASH_MSG = (By.CSS_SELECTOR, "[data-testid='flash-message']")
    SWITCH_LINK = (By.CSS_SELECTOR, "[data-testid='auth-switch-link']")

    def open_login(self) -> "AuthPage":
        """Navigate directly to the Login page."""
        self.driver.get(f"{self.base_url}/login")
        self.find(*self.TITLE)
        return self

    def open_register(self) -> "AuthPage":
        """Navigate directly to the Register page."""
        self.driver.get(f"{self.base_url}/register")
        self.find(*self.TITLE)
        return self

    def get_title_text(self) -> str:
        """Get the h1 title on the auth form."""
        return self.get_text(*self.TITLE)

    def fill_email(self, email: str) -> "AuthPage":
        """Fill email input field."""
        self.type_text(*self.EMAIL_INPUT, email)
        return self

    def fill_password(self, password: str) -> "AuthPage":
        """Fill password input field."""
        self.type_text(*self.PASSWORD_INPUT, password)
        return self

    def fill_confirm_password(self, confirm_password: str) -> "AuthPage":
        """Fill confirm password input field (register mode)."""
        self.type_text(*self.CONFIRM_PASSWORD_INPUT, confirm_password)
        return self

    def submit(self) -> None:
        """Click the submit button."""
        self.click(*self.SUBMIT_BTN)

    def click_google_login(self) -> None:
        """Click the Google OAuth login button."""
        self.click(*self.GOOGLE_BTN)

    def login(self, email: str, password: str) -> None:
        """Convenience method to fill and submit the login form."""
        self.fill_email(email)
        self.fill_password(password)
        self.submit()

    def register(self, email: str, password: str = "Password123!", confirm_password: str = None) -> None:
        """Convenience method to fill and submit the registration form."""
        user_confirm = confirm_password if confirm_password is not None else password

        self.fill_email(email)
        self.fill_password(password)
        self.fill_confirm_password(user_confirm)
        self.submit()

    def get_flash_message(self, timeout: int = 5) -> str:
        """Get text of the flash error or success message."""
        return self.get_text(*self.FLASH_MSG, timeout=timeout)

    def has_flash_message(self, timeout: int = 2) -> bool:
        """Check if a flash message is present on the page."""
        return self.is_visible(*self.FLASH_MSG, timeout=timeout)

    def click_switch_link(self) -> None:
        """Click the toggle link to switch between Login and Register."""
        self.click(*self.SWITCH_LINK)

    def is_at_login(self, timeout: int = 5) -> bool:
        """Check if currently on the Login page."""
        try:
            self.wait_for(lambda d: "/login" in d.current_url, timeout=timeout)
            return True
        except Exception:
            return False

    def is_at_register(self, timeout: int = 5) -> bool:
        """Check if currently on the Register page."""
        try:
            self.wait_for(lambda d: "/register" in d.current_url, timeout=timeout)
            return True
        except Exception:
            return False
